control_panel_api/authentication.py

```diff
@@ -1,54 +1,94 @@
+import logging
+
 from django.conf import settings
-import jwt
-from jwt.exceptions import InvalidTokenError
+from jose import jwt
+from jose.exceptions import JWTError
+import requests
+from requests.exceptions import RequestException
 from rest_framework.authentication import (
     BaseAuthentication,
-    get_authorization_header
+    get_authorization_header,
 )
 from rest_framework.exceptions import AuthenticationFailed
 
 from control_panel_api.models import User
 
+logger = logging.getLogger(__name__)
 
-class Auth0JWTAuthentication(BaseAuthentication):
 
-    def authenticate(self, request):
+def get_jwks():
+    response = requests.get(settings.OIDC_WELL_KNOWN_URL)
+    return response.json()
+
+
+def get_key(jwks, kid):
+    for jwk in jwks.get('keys'):
+        if jwk['kid'] == kid:
+            return jwk
+
+    return None
+
+
+def get_jwk(kid):
+    jwks = get_jwks()
+
+    return get_key(jwks, kid)
+
+
+def get_or_create_user(payload):
```

Comment:
I'd probably rename `payload` => `token`.
Also, shouldn't this be a "private" method of `Auth0JWTAuthentication`?

Reply:
It's the decoded payload.  The token is the encrypted string passed in the header.  After decoding it's split into the payload.  The token also contains a header and secret fyi.

Yes those functions can be part of the class. Not sure why I left them out.

Reply:
You're right, I suppose in my mind when I think "token" I keep thinking about the decoded JSON but the token is the string getting in. Still `payload` is a bit generic, maybe something like `decoded_token` or something like that would work better...but nevermind, it was just an idea.